### DISCUSSION TITLE
Pick colour not working for PNG, but works for JPG

### DISCUSSION THREAD

**Discussion by MaxiAxi-Jorn** [ImageMagick/ImageMagick#7170] [Help] [answered]
### ImageMagick version

6.9.11-60

### Operating system, version and so on

Debian Buster

### Description

Hi,

I'm trying to resize my rectangle images into square images.
Can be vertical en horizontal rectangle. I want to extend those images with a white background.
Most images are product shots with a white background. But some images are lifestyle shots, so I don't want to resize those (since cropping them into square is better).

So I want to check if the upper left is white (yes = product image, no = lifestyle image). If white, then extend image with white background. If not white, skip and log file (so can be done later manually).

I've created this script and it's processing JPG images just fine, but not PNG images. It doesn't recognise the white background (and they're actually white, not transparant).

This is the script:
```
#!/bin/bash

# Log file for skipped images
LOG_FILE="skipped_images.log"

# Function to check if a color is white
is_white_color() {
    local color="$1"
    [[ "$color" =~ ^#?(FFFFFF|fff|white)$ ]]
}

# Function to process images
process_image() {
    local file="$1"
    local dimensions=$(identify -format "%w %h" "$file")
    read -r width height <<< "$dimensions"
    if (( width != height )); then
        # Get the color of the upper left corner of the image
        corner_color=$(convert "$file" -depth 8 -crop 1x1+0+0 txt:- | grep -oP '(?<=\#)[0-9a-fA-F]+')
        if is_white_color "$corner_color"; then
            max_side=$(( width > height ? width : height ))
            convert "$file" -gravity center -background white -extent "${max_side}x${max_side}" "$file"
            echo "Processed $file"
        else
            echo "Skipped $file (upper left corner color is not white)"
            echo "$file" >> "$LOG_FILE"
        fi
    else
        echo "Skipped $file (already square)"
    fi
}

# Clear the log file
> "$LOG_FILE"

# Iterate over JPEG and PNG image files in the current directory and its subdirectories
find . -type f \( -iname "*.jpg" -o -iname "*.jpeg" -o -iname "*.jpe" -o -iname "*.png" \) | while read -r file; do
    process_image "$file"
done
```

And I've attached the images.

Does anyone know why it's not processing the PNG files correctly?
![rectangle_vertical_before](https://github.com/ImageMagick/ImageMagick/assets/72382701/921da002-132e-4a5e-9215-f297936b2ebe)
<img width="181" alt="rectangle_vertical_before" src="https://github.com/ImageMagick/ImageMagick/assets/72382701/a97e7287-ec26-45c8-bef2-75c8bb9ac17d">

**Comment by fmw42** [comment]
Your PNG file has an opaque alpha channel.  Your test for white does not include alpha values.  If you believe that all your images should be opaque, then after reading the image add -alpha off and +repage.

Otherwise, add tests for opaque alpha in your check for white.

**Reply by MaxiAxi-Jorn** [reply]
Thank you, I'll check this out!

**Reply by fmw42** [reply, marked as answer]
Sorry, your PNG alpha  has strange values (infinite).  How was this PNG created?

```
    Alpha:
      min: inf  (2.7431e+303)
      max: -inf (-2.7431e+303)
 
```


However, a test of the colors at the 0,0 (top-left) pixel shows and opaque white pixel.:

magick img.png -format "%[pixel:u.p{0,0}]" info:
srgba(255,255,255,1)

or in hex as

```
magick img.png -format "%[hex:u.p{0,0}]" info:
FFFFFFFF

```

The hex color might be easier to test:  either FFFFFF or FFFFFFFF, rather than all your options.

**Reply by MaxiAxi-Jorn** [reply]
This example images were created in MS Paint. The actual images have different sources, so it can be any kind of PNG.
Since I'm already testing for FFFFFF, should I only add FFFFFFFF?

**Reply by fmw42** [reply]
That should work. Give it a try

**Reply by fmw42** [reply]
How do you get "local color"?  Can you ensure that it is not formatted as rgb(rr,gg,bb)?

**Reply by MaxiAxi-Jorn** [reply]
This seems to work indeed! I'm now going to test it with a larger set of images.

**Comment by MaxiAxi-Jorn** [comment]
> Sorry, your PNG alpha has strange values (infinite). How was this PNG created?
> 
> ```
>     Alpha:
>       min: inf  (2.7431e+303)
>       max: -inf (-2.7431e+303)
>  
> ```
> 
> However, a test of the colors at the 0,0 (top-left) pixel shows and opaque white pixel.:
> 
> magick img.png -format "%[pixel:u.p{0,0}]" info: srgba(255,255,255,1)
> 
> or in hex as
> 
> ```
> magick img.png -format "%[hex:u.p{0,0}]" info:
> FFFFFFFF
> ```
> 
> The hex color might be easier to test: either FFFFFF or FFFFFFFF, rather than all your options.



> That should work. Give it a try
